File: ployglots/mta-ployglot-pkg-figma-api/packages/py/figma_api/_figma_client.py
"""FigmaClient — the top-level async SDK entry point."""

from __future__ import annotations

from typing import Any

from ._config import FigmaConfig, resolve_figma_config
from ._errors import FigmaError, FigmaErrorContext, map_http_error
from ._fetch_client import FetchClient, create_default_fetch_client
from ._logger import Logger, create_logger
from .resources.comments import CommentsResource
from .resources.components import ComponentsResource
from .resources.dev_resources import DevResourcesResource
from .resources.files import FilesResource
from .resources.library_analytics import LibraryAnalyticsResource
from .resources.me import MeResource
from .resources.projects import ProjectsResource
from .resources.variables import VariablesResource
from .resources.webhooks import WebhooksResource


class FigmaClient:
    """Async Figma API client.

    Usage::

        async with FigmaClient(proxy={}) as client:
            me = await client.me.get()
    """

    def __init__(
        self,
        *,
        token: str | None = None,
        host: str | None = None,
        user: str | None = None,
        proxy: dict[str, Any] | None = None,
        timeout_ms: int = 30_000,
        default_headers: dict[str, str] | None = None,
        retry: dict[str, Any] | bool | None = None,
        force_overwrite_retry: bool = False,
        fetch_client: FetchClient | None = None,
        logger: Logger | None = None,
    ) -> None:
        self.config: FigmaConfig = resolve_figma_config(
            token=token,
            host=host,
            user=user,
            proxy=proxy,
            timeout_ms=timeout_ms,
            default_headers=default_headers,
            retry=retry,
            force_overwrite_retry=force_overwrite_retry,
        )
        self.fetch_client: FetchClient = fetch_client or create_default_fetch_client(self.config)
        self.logger: Logger = logger or create_logger(prefix="figma-api")

        self.me = MeResource(self)
        self.files = FilesResource(self)
        self.comments = CommentsResource(self)
        self.projects = ProjectsResource(self)
        self.components = ComponentsResource(self)
        self.variables = VariablesResource(self)
        self.dev_resources = DevResourcesResource(self)
        self.library_analytics = LibraryAnalyticsResource(self)
        self.webhooks = WebhooksResource(self)

    async def get(self, path: str, **kwargs: Any) -> Any:
        return await self._request("GET", path, **kwargs)

    async def post(self, path: str, **kwargs: Any) -> Any:
        return await self._request("POST", path, **kwargs)

    async def delete(self, path: str, allow_empty: bool = True, **kwargs: Any) -> Any:
        return await self._request("DELETE", path, allow_empty=allow_empty, **kwargs)

    async def _request(
        self, method: str, path: str, *, allow_empty: bool = False, **kwargs: Any
    ) -> Any:
        verb = getattr(self.fetch_client, method.lower())
        resp = await verb(path, **kwargs)
        status = int(getattr(resp, "status_code", getattr(resp, "status", 0)))
        if status >= 400:
            body = ""
            try:
                body = getattr(resp, "text", "") or ""
                if callable(body):
                    body = body()
            except Exception:
                body = ""
            retry_after = None
            try:
                retry_after = resp.headers.get("retry-after")
            except Exception:
                retry_after = None
            raise map_http_error(
                FigmaErrorContext(
                    method=method,
                    url=path,
                    status=status,
                    body=str(body or ""),
                    retry_after=retry_after,
                )
            )
        try:
            return resp.json()
        except Exception as err:
            if allow_empty:
                return None
            raise FigmaError(f"{method} {path} → failed to decode JSON") from err

    async def aclose(self) -> None:
        close = getattr(self.fetch_client, "aclose", None)
        if close is not None:
            await close()

    async def __aenter__(self) -> FigmaClient:
        return self

    async def __aexit__(self, exc_type: object, exc: object, tb: object) -> None:
        await self.aclose()

File: ployglots/mta-ployglot-pkg-figma-api/packages/ts/src/figma-client.ts
// @ts-nocheck
/**
 * `FigmaClient` — the top-level SDK entry point.
 *
 * Composes the FetchClient contract with domain sub-SDKs. Accepts
 * either the default (token + proxy + env-driven) fetch client, or a
 * BYO one produced by `fetchClientFromPolyglot(outer)` for composed
 * retry / caching / circuit-breaker topologies.
 */

import { resolveFigmaConfig, type FigmaConfig, type FigmaConfigInput } from './config.js';
import { FigmaError, mapHttpError } from './errors.js';
import {
  createDefaultFetchClient,
  type FetchClient,
  type FetchRequestOptions,
} from './fetch-client.js';
import { createLogger, type Logger } from './logger.js';
import { CommentsResource } from './resources/comments.js';
import { ComponentsResource } from './resources/components.js';
import { DevResourcesResource } from './resources/dev-resources.js';
import { FilesResource } from './resources/files.js';
import { LibraryAnalyticsResource } from './resources/library-analytics.js';
import { MeResource } from './resources/me.js';
import { ProjectsResource } from './resources/projects.js';
import { VariablesResource } from './resources/variables.js';
import { WebhooksResource } from './resources/webhooks.js';

export interface FigmaClientOptions extends FigmaConfigInput {
  /** Pre-built FetchClient to bypass the default transport. */
  fetchClient?: FetchClient;
  logger?: Logger;
}

export class FigmaClient {
  readonly config: FigmaConfig;
  readonly fetchClient: FetchClient;
  readonly logger: Logger;

  readonly me: MeResource;
  readonly files: FilesResource;
  readonly comments: CommentsResource;
  readonly projects: ProjectsResource;
  readonly components: ComponentsResource;
  readonly variables: VariablesResource;
  readonly devResources: DevResourcesResource;
  readonly libraryAnalytics: LibraryAnalyticsResource;
  readonly webhooks: WebhooksResource;

  constructor(options: FigmaClientOptions = {}) {
    this.config = resolveFigmaConfig(options);
    this.fetchClient = options.fetchClient ?? createDefaultFetchClient(this.config);
    this.logger = options.logger ?? createLogger({ prefix: 'figma-api' });

    this.me = new MeResource(this);
    this.files = new FilesResource(this);
    this.comments = new CommentsResource(this);
    this.projects = new ProjectsResource(this);
    this.components = new ComponentsResource(this);
    this.variables = new VariablesResource(this);
    this.devResources = new DevResourcesResource(this);
    this.libraryAnalytics = new LibraryAnalyticsResource(this);
    this.webhooks = new WebhooksResource(this);
  }

  /**
   * Low-level GET helper used by every resource. Centralises error
   * mapping so subclasses stay free of HTTP concerns.
   */
  async get<T>(path: string, options: FetchRequestOptions = {}): Promise<T> {
    return this._request<T>('GET', path, options);
  }

  async post<T>(path: string, options: FetchRequestOptions = {}): Promise<T> {
    return this._request<T>('POST', path, options);
  }

  async delete<T>(path: string, options: FetchRequestOptions = {}): Promise<T> {
    return this._request<T>('DELETE', path, options, true);
  }

  private async _request<T>(
    method: 'GET' | 'POST' | 'DELETE',
    path: string,
    options: FetchRequestOptions,
    allowEmpty = false,
  ): Promise<T> {
    const verb =
      method === 'GET'
        ? this.fetchClient.get.bind(this.fetchClient)
        : method === 'POST'
          ? this.fetchClient.post.bind(this.fetchClient)
          : this.fetchClient.delete.bind(this.fetchClient);
    const resp = await verb(path, options);
    const status = statusOf(resp);
    if (status >= 400) {
      const body = await safeText(resp);
      throw mapHttpError({
        method,
        url: path,
        status,
        body,
        retryAfter: headerOf(resp, 'retry-after'),
      });
    }
    try {
      return (await resp.json()) as T;
    } catch (err) {
      if (allowEmpty) return undefined as unknown as T;
      throw new FigmaError(`${method} ${path} → failed to decode JSON`, { cause: err });
    }
  }

  async close(): Promise<void> {
    await this.fetchClient.close();
  }
}

async function safeText(resp: { text?: () => Promise<string> }): Promise<string> {
  try {
    return (await resp.text?.()) ?? '';
  } catch {
    return '';
  }
}

/**
 * Read the HTTP status off a response. Handles both the fetch-http-client
 * `Response` (`.statusCode`) and the fetch-style `.status` getter — the
 * fake test client uses the latter and we want both to work.
 */
function statusOf(resp: unknown): number {
  const r = resp as { statusCode?: number; status?: number };
  return r.statusCode ?? r.status ?? 0;
}

function headerOf(resp: unknown, name: string): string | null {
  const r = resp as { headers?: { get?: (n: string) => string | null | undefined } };
  return r.headers?.get?.(name) ?? null;
}
